processing/src/main/java/io/druid/query/extraction/SubstringDimExtractionFn.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.extraction;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+/**
+ */
+public class SubstringDimExtractionFn extends DimExtractionFn
+{
+  private static final byte CACHE_TYPE_ID = 0x8;
+
+  private final int index;
+  private final int length;
+  private final int end;
+
+  @JsonCreator
+  public SubstringDimExtractionFn(
+      @JsonProperty("index") int index,
+      @Nullable
+      @JsonProperty("length") Integer length
+  )
+  {
+
+    Preconditions.checkArgument(length == null || length >= 0, "length must be positive");
+
+    this.index = index;
+    this.length = length != null ? length : -1;
+    this.end = index + this.length;
```

Comment:
True, I am still learning my way around and wasn't sure how often getLength, equals, and hashCode would be called, though the latter 2 could be refactored to use 'end' instead if we were to dump length. Basically, is it worse to store an extra integer or have to perform arithmetic every time getLength() is called? I added end in the first place because I didn't want to have to calculate it on every call to apply().